python/tskit/combinatorics.py

```diff
@@ -28,12 +28,133 @@
 import functools
 import heapq
 import itertools
+import json
+import random
 
+import attr
 import numpy as np
 
 import tskit
 
 
+@attr.s(eq=False)
+class TreeNode:
+    """
+    Simple linked tree class used to generate tree topologies.
+    """
+
+    parent = attr.ib(default=None)
+    children = attr.ib(factory=list)
+    label = attr.ib(default=None)
+
+    @staticmethod
+    def random_binary_tree(leaf_labels, rng):
+        """
+        Based on the description of Remy's method of generating "decorated"
+        random binary trees in TAOCP 7.2.1.6. This is not a direct
+        implementation of Algorithm R, because we are interested in
+        the leaf node labellings.
+
+        The pre-fasicle text is available here, page 16:
```

Comment:
I've never heard of a fasicle?! Looks like it has an extra c in the spelling anyway!